lib/ContentTypeParser.js

```diff
@@ -1,71 +1,184 @@
 
   if (this.hasParser(contentType)) {
     throw new Error(`Content type parser '${contentType}' already present.`)
   }
+
+  parser.asString = opts.asString === true
+  parser.asBuffer = opts.asBuffer === true
+
   if (contentType === '*') {
     this.parserList.push('')
-    this.customParsers[''] = fn
+    this.customParsers[''] = parser
   } else {
-    this.parserList.unshift(contentType)
-    this.customParsers[contentType] = fn
+    if (contentType !== 'application/json') {
+      this.parserList.unshift(contentType)
+    }
+    this.customParsers[contentType] = parser
   }
 }
 
 ContentTypeParser.prototype.hasParser = function (contentType) {
+  if (contentType === 'application/json') return false
   return contentType in this.customParsers
 }
 
 ContentTypeParser.prototype.fastHasHeader = function (contentType) {
-  if (!contentType) return this.hasParser('')
-
   for (var i = 0; i < this.parserList.length; i++) {
     if (contentType.indexOf(this.parserList[i]) > -1) return true
   }
   return false
 }
 
-ContentTypeParser.prototype.getHandler = function (contentType) {
-  if (contentType === undefined) return this.customParsers['']
-
+ContentTypeParser.prototype.getParser = function (contentType) {
   for (var i = 0; i < this.parserList.length; i++) {
     if (contentType.indexOf(this.parserList[i]) > -1) {
-      return this.customParsers[this.parserList[i]]
+      var parser = this.customParsers[this.parserList[i]]
+      this.cachedParsers[contentType] = parser
+      return parser
     }
   }
+
+  return this.customParsers['']
 }
 
 ContentTypeParser.prototype.run = function (contentType, handler, request, reply) {
-  var result = this.getHandler(contentType)(request.raw, done)
-  if (result && typeof result.then === 'function') {
-    result
-      .then(body => done(null, body))
-      .catch(done)
+  var parser = this.cachedParsers[contentType] || this.getParser(contentType)
+  if (parser === undefined) {
+    reply.code(415).send(new Error('Unsupported Media Type: ' + contentType))
+  } else {
+    if (parser.asString === true || parser.asBuffer === true) {
+      rawBody(
+        request,
+        reply,
+        reply.context._parserOptions,
+        parser,
+        handler
+      )
+    } else {
+      var result = parser(request.raw, done)
+      if (result && typeof result.then === 'function') {
+        result.then(body => done(null, body), done)
+      }
+    }
   }
 
   function done (error, body) {
     if (error) {
-      return reply.send(error)
+      reply.send(error)
+    } else {
+      request.body = body
+      handler(reply)
+    }
+  }
+}
+
+function rawBody (request, reply, options, parser, handler) {
+  var asString = parser.asString
+  var limit = options.limit
+  var contentLength = request.headers['content-length'] === undefined
+    ? NaN
+    : Number.parseInt(request.headers['content-length'], 10)
+
+  if (contentLength > limit) {
+    reply.code(413).send(new Error('Request body is too large'))
```

Comment:
Since the `Error` is duplicated in the `onData` handler, I think it would be wise to use a variable for it. Thus, if the message needs to be updated for some reason, it only has to be updated in one place.

Reply:
Correct me if I'm wrong, but if we cache the error object the stack trace will not be super easy to read.

Reply:
I'm not sure. Could just cache the string.